src/activity_type.cpp

```diff
@@ -104,3 +132,19 @@ void activity_type::reset()
     activity_type_all.clear();
 }
 
+void activity_type::spawn_do_turn_byproducts( player *p ) const
+{
+    if( do_turn_byproducts_item_group && calendar::once_every( do_turn_byproducts_interval ) ) {
+        g->m.spawn_items( p->pos(), item_group::items_from( *do_turn_byproducts_item_group,
```

Comment:
This puts the pile of excavated materials right under the player's feet. I'd suggest choosing an adjacent tile (a passable tile which isn't the pit itself) or let the player specify the location before start digging.

Reply:
Yes it does. Can you expand a little bit on why that's problematic? I don't necessarily have a problem with changing it but I think:
* the generic byproduct spawning won't always make sense to put in another adjacent location, so if I do it in another prompted location I'd probably hoist the prompt into the iuse that triggers the activity and the spawn the byproducts out of the activity itself rather than the generic version
* there won't always be a passable tile which isn't the pit and isn't the players location

Reply:
My thought on the matter are:
- In ASCII mode the pile will be totally obscured by the character. You won't see it untill you step aside.
- Handling of loose materials is a probable subject for a change: moving piles of powder around should be more time consuming than picking up/dropping solid items (probably far more). So you might want to put the excavated materials on the right tile (or in a wheelbarrow) in one go.
- Most of the constructions require a tile free of any items. It's not the case for a pit: you can dig it under/around anything (event under the soil that you've just excavated from another pit of the same dimensions). That seems to be a bit inconsistent.
- Digging IRL usually implies placing the excavated soil to a nearby spot, not right under one's feet.

> I'd probably hoist the prompt into the iuse that triggers the activity and the spawn the byproducts out of the activity itself rather than the generic version

I don't quite see why is that problematic implementation-wise. You can still spawn the byproduct during the activity no matter the prompt.

> there won't always be a passable tile which isn't the pit and isn't the players location

Player's location can be the last resort in this case.

Reply:
> Handling of loose materials is a probable subject for a change: moving piles of powder around should be more time consuming than picking up/dropping solid items (probably far more). So you might want to put the excavated materials on the right tile (or in a wheelbarrow) in one go.

Doesn't seem unreasonable, though the [quantity of material I ballparked this with](https://github.com/CleverRaven/Cataclysm-DDA/pull/28942/files#diff-74bb173f819041f82ce1da375b45569cR2495) would be more than a wheelbarrow's capacity, but the general case of wanting to not do the work twice makes sense. 

> Most of the constructions require a tile free of any items. It's not the case for a pit: you can dig it under/around anything (event under the soil that you've just excavated from another pit of the same dimensions). That seems to be a bit inconsistent.

Changing that is a bit of a scope creep from what I was going for here (adding byproducts to existing digging behavior), but I agree that it'd make more sense to require the tile be clear. Seems like a good follow-up change.

> Digging IRL usually implies placing the excavated soil to a nearby spot, not right under one's feet.

Digging an area of the sizes implied here to any significant depth by hand is going to involve standing on/in the location being dug and tossing material out. I've just imagined the character is standing on their pile of excavated material after finishing the task.

> I don't quite see why is that problematic implementation-wise. You can still spawn the byproduct during the activity no matter the prompt.

I think I was probably thinking about where I thought the prompt and validation of the result would have to happen (digging iuse, stuffed into activity coords) and how it'd have to get utilized (pulled out of activity coords in the associated activity do_turn or finish) since the generic spawn in `activity_type::call_finish` wouldn't know how to interpret activity specific data. Where were you thinking those various steps would happen?

If anything, these comments imply that we should probably update the dig activity such that:
1. you can only dig on your current location (currently you can dig any adjacent tile, but not your current location)
2. that it must be free of items
3. that you must pick an adjacent passable location for the byproducts, and if no such valid location exists, you can't dig

Reply:
> I think I was probably thinking about where I thought the prompt and validation of the result would have to happen (digging iuse, stuffed into activity coords) and how it'd have to get utilized (pulled out of activity coords in the associated activity do_turn or finish)

Sounds about right: the prompt in `use::dig` and its result in `player_activity::coords`. This approach seems to be conventional (though not ideal, since `player_activity` tries to suite all needs and therefore it's polluted with unrelated data).

> Where were you thinking those various steps would happen?

I think the concept of `spawn_do_turn_byproducts`/`spawn_finish_byproducts` is flawed to begin with. The intermediate/final outcome is way too activity-dependent, so the handlers should probably decide for themselves whether they spawn anything at any stage and [what](https://github.com/CleverRaven/Cataclysm-DDA/pull/28942#issuecomment-476137699) it would be. Furthermore, all the activities (except one) don't spawn anything each turn, hence they don't need this specific knowledge + data in `activity_type` (I mean I'd leave `activity_type` as abstract as it can be).

>  1. you can only dig on your current location (currently you can dig any adjacent tile, but not your current location)
> 2. that it must be free of items
> 3. that you must pick an adjacent passable location for the byproducts, and if no such valid location exists, you can't dig

This looks perfectly reasonable to me.